### PR TITLE
More HIP fixes

### PR DESCRIPTION
**Pull request by rbberger**
1. Some test files weren't being set to LANGUAGE HIP. You also need to set the linker language to HIP to avoid errors due to unknown compiler / linker flags.
2. On systems with older GCCs as system default, amdclang will fail to compile C++17 or newer code that uses C++ standard library features that only exist in that C++ standard. The standard library headers and library must match that standard. This adds logic to detect if `CMAKE_CXX_COMPILER` is set to a GNU compiler and propagates its location to amdclang via `--gcc-install-dir` and an path to the standard library location.

### CODE REVIEW

File: cmake/macros.cmake [hunk 1]
```diff
@@ -1296,12 +1298,13 @@ FUNCTION( FINALIZE_TESTBUILDER )
     ENDIF()
     # If we are using HIP and COMPILE_CXX_AS_HIP is set, change the language
     IF ( USE_HIP AND COMPILE_CXX_AS_HIP )
-        FOREACH ( tmp ${TESTBUILDER_SOURCES} )
+        FOREACH ( tmp ${TB_TARGET_SOURCES} )
```
Comment (bobby-philip) [inline cmake/macros.cmake:1301]: @rbberger, Could lines 1278-279 be deleted and 1300 be modified to be 

 IF ( TESTBUILDER_SOURCES AND USE_HIP AND COMPILE_CXX_AS_HIP )

File: cmake/macros.cmake [hunk 2]
```diff
@@ -1296,12 +1298,13 @@ FUNCTION( FINALIZE_TESTBUILDER )
...
             GET_SOURCE_FILE_PROPERTY( lang "${tmp}" LANGUAGE ) 
-            IF ( lang STREQUAL "CXX" )
+            IF ( lang STREQUAL "CXX" OR lang STREQUAL "NOTFOUND" )
```
Comment (bobby-philip) [inline cmake/macros.cmake:1303]: @rbberger , code looks fine, just curious how lang ended up not being found

Comment (rbberger) [reply cmake/macros.cmake:1303]: The source files are added in cmake without a file extension. So there is no heuristic CMake can follow to determine the language. The name is further used in the [code generation](https://github.com/AdvancedMultiPhysics/TPL-builder/blob/master/cmake/macros.cmake#L1312) portion without extension.

File: cmake/macros.cmake (hunk 1 above)
Comment (rbberger) [reply cmake/macros.cmake:1301]: No. The problem is when `TESTBUILDER_SOURCES` is empty, nothing sets the necessary HIP language for `TestBuilder.cpp`, causing linker errors.

Comment (rbberger) [reply cmake/macros.cmake:1301]: Oh wait, in that case also the linker language stays unmodified. Let me try.

File: cmake/macros.cmake (hunk 2 above)
Comment (rbberger) [reply cmake/macros.cmake:1303]: So the alternative is to make that builder "smarter" by accepting files with extension and then stripping it when needed.

File: cmake/macros.cmake (hunk 1 above)
Comment (rbberger) [reply cmake/macros.cmake:1301]: Nope, the flags are injected by Kokkos. So unless you remove the kokkos dependence in that case too, you will get that error.
```
g++: error: unrecognized command-line option '-fno-gpu-rdc'
g++: error: unrecognized command-line option '--offload-arch=gfx90a'
```

File: cmake/macros.cmake (hunk 2 above)
Comment (rbberger) [reply cmake/macros.cmake:1303]: @bobby-philip pushed a new version that uses `gcc -print-file-name`, which seems much more robust.

File: cmake/macros.cmake (hunk 1 above)
Comment (bobby-philip) [reply cmake/macros.cmake:1301]: Hmm -- I think the intent was that in the case TESTBUILDER_SOURCES is empty TestBuilder.cpp is not compiled. But I guess you are saying that it is?

Comment (rbberger) [reply cmake/macros.cmake:1301]: Yes, see https://github.com/AdvancedMultiPhysics/AMP/blob/master/src/operators/test/CMakeLists.txt#L150-L179. We're only building with `+mpi+kokkos+rocm`.

Comment (rbberger) [reply cmake/macros.cmake:1301]: The executable is always added during [INITIALIZE_TESTBUILDER](https://github.com/AdvancedMultiPhysics/TPL-builder/blob/master/cmake/macros.cmake#L1248-L1252). To get the behavior you outlined above, you would need to move to FINALIZE_TESTBUILDER.

Comment (bobby-philip) [reply cmake/macros.cmake:1301]: Okay, sounds good. Mark will change if he wants that behavior

**Pull request merged**